cmd/homi/setup/cmd.go

```diff
@@ -786,6 +810,66 @@ func makeSCNodes(num int, isWorkOnSingleHost bool) ([]*ValidatorInfo, []string)
 	return scn, scnNodeKeys
 }
 
+func makeSPNNodes(num int, isWorkOnSingleHost bool) ([]*ValidatorInfo, []string) {
+	privKeys, nodeKeys, nodeAddrs := istcommon.GenerateKeys(num)
+
+	var p2pPort uint16
+	var proxies []*ValidatorInfo
+	var proxyNodeKeys []string
+	for i := 0; i < num; i++ {
+		if isWorkOnSingleHost {
+			p2pPort = lastIssuedPortNum
+			lastIssuedPortNum++
+		} else {
+			p2pPort = DefaultTcpPort
+		}
+
+		v := &ValidatorInfo{
+			Address: nodeAddrs[i],
+			Nodekey: nodeKeys[i],
+			NodeInfo: discover.NewNode(
+				discover.PubkeyID(&privKeys[i].PublicKey),
+				net.ParseIP("0.0.0.0"),
+				0,
+				p2pPort,
+				discover.NodeTypeUnknown).String(),
+		}
+		proxies = append(proxies, v)
+		proxyNodeKeys = append(proxyNodeKeys, v.Nodekey)
+	}
+	return proxies, proxyNodeKeys
+}
+
+func makeSENNodes(num int, isWorkOnSingleHost bool) ([]*ValidatorInfo, []string) {
```

Comment:
makeSPNNodes and makeSENNodes seem to do exactly the same thing. Do I need to repeat the same code?

Reply:
Fixed.